plutus-scb/src/Cardano/Metadata/Types.hs

```diff
@@ -66,67 +82,148 @@ newtype PropertyName =
+             pure $ Name value signatures)
+
+parseDescription :: JSON.Object -> Parser (PropertyDescription 'ExternalEncoding)
+parseDescription =
+    parseAtKey
+        "description"
+        (\description -> do
+             value <- description .: "value"
+             signatures <- description .: "anSignatures"
+             pure $ Description value signatures)
+
+parsePreimage :: JSON.Object -> Parser (PropertyDescription 'ExternalEncoding)
+parsePreimage =
+    parseAtKey
+        "preImage"
+        (\preImage -> do
+             hash <- preImage .: "hashFn"
+             hex <- preImage .: "hex"
+             pure $ Preimage hash hex)
+
+parseOwner :: JSON.Object -> Parser (PropertyDescription 'ExternalEncoding)
+parseOwner subObject = do
+    sig <- subObject .: "owner"
+    pure $ Other "owner" (JSON.Object subObject) (pure sig)
+
+parseAtKey ::
+       Text -> (JSON.Object -> Parser a) -> JSON.Object -> Parser a
+parseAtKey key parser o = do
+    subValue <- o .: key
+    withObject (Text.unpack key) parser subValue
+
+data HashFunction
+    = SHA256
+    | Blake2B256
+    deriving (Show, Eq, Ord, Generic, Enum, Bounded)
 
 instance FromJSON HashFunction where
     parseJSON =
         withText "HashFunction" $ \case
             "SHA256" -> pure SHA256
+            "blake2b-256" -> pure Blake2B256
             other -> fail $ "Unknown HashFunction '" <> Text.unpack other <> "'"
 
 instance ToJSON HashFunction where
-    toJSON = JSON.String . tshow
-
-data PropertyDescription
-    = Preimage HashFunction LedgerBytes
-    | Name Text (NonEmpty AnnotatedSignature)
-    | Description Text (NonEmpty AnnotatedSignature)
-    | Other Text JSON.Value (NonEmpty AnnotatedSignature)
-    deriving (Show, Eq, Generic)
-    deriving anyclass (ToJSON, FromJSON)
+    toJSON SHA256     = JSON.String "SHA256"
+    toJSON Blake2B256 = JSON.String "blake2b-256"
 
-data AnnotatedSignature =
+data AnnotatedSignature (encoding :: k) =
     AnnotatedSignature PubKey Signature
     deriving (Show, Eq, Ord, Generic)
-    deriving anyclass (ToJSON, FromJSON)
 
-makeLenses ''Property
+deriving anyclass instance ToJSON (AnnotatedSignature 'AesonEncoding)
+deriving anyclass instance FromJSON (AnnotatedSignature 'AesonEncoding)
 
-toId :: Property -> (Subject, PropertyKey)
-toId (Property subject description) = (subject, toPropertyKey description)
+instance FromJSON (AnnotatedSignature 'ExternalEncoding) where
+    parseJSON =
+        withObject "AnnotatedSignature" $ \o -> do
+            sigRaw <- o .: "signature"
+            sigBytes <- decodeByteString sigRaw
+            let sig = Signature sigBytes
+            pubKeyRaw :: Text <- o .: "publicKey"
+            case PubKey <$> LedgerBytes.fromHex (encodeUtf8 pubKeyRaw) of
+                Right pubKey -> pure $ AnnotatedSignature pubKey sig
+                Left err     -> fail (show (err, pubKeyRaw))
+
+data JSONEncoding = ExternalEncoding | AesonEncoding
+
+data PropertyDescription (encoding :: k)
```

Comment:
I don't get this name, this suggests a description but it is in fact a property or a piece of data?

Reply:
I was going to say, "this is from the spec." But I just double-checked the spec and it shrugged at me.

Changed to `Property`. Thanks. :-)